### PR TITLE
Converting to async - Initial flow

### PR DESCRIPTION
**Pull request by avik-taranis**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+FROM python:3.12-slim
+
+ARG BUILD_LOCAL
+
+COPY requirements.txt test_requirements.txt setup.py README.md ./
```
Comment (razsinay) [inline Dockerfile:5]: Why do we need to copy the README.md?

File: py_async_grpc_prometheus/prometheus_async_server_interceptor.py [hunk 2]
```diff
@@ -118,30 +97,81 @@ def new_behavior(request_or_iterator, servicer_context):
...
+            if request_streaming:
+              request_or_iterator = grpc_utils.wrap_iterator_inc_counter(
+                  request_or_iterator,
+                  self._metrics["grpc_server_stream_msg_received"],
+                  grpc_type,
+                  grpc_service_name,
+                  grpc_method_name)
 
-    optional_any = self._wrap_rpc_behavior(continuation(handler_call_details), metrics_wrapper)
+            async for obj in behavior(request_or_iterator, servicer_context):
```
Comment (razsinay) [inline py_async_grpc_prometheus/prometheus_async_server_interceptor.py]: I suggest to find a better name for `obj`

File: py_async_grpc_prometheus/prometheus_async_server_interceptor.py [hunk 3]
```diff
@@ -118,30 +97,81 @@ def new_behavior(request_or_iterator, servicer_context):
...
+        
+        except Exception as e:
+          if self._skip_exceptions:
+            if self._log_exceptions:
+              _LOGGER.error(e)
+            async for obj in behavior(request_or_iterator, servicer_context):
+              yield obj
+          raise e
+      
+      if response_streaming:
```
Comment (razsinay) [inline py_async_grpc_prometheus/prometheus_async_server_interceptor.py:154]: I am a bit confused with the multiple `try/except` blocks.
 Maybe we can wrap some of the parts in functions?

**Review by razsinay — APPROVED** [review]
I think it is worth adding so documentation to the functions, maybe describe inside the function the general flow, as this code is not trivial.

File: Dockerfile (hunk 1 above)
Comment (avik-taranis) [reply Dockerfile:5]: Due to the long description in setup.py:

`with open("README.md", "r") as fh:
  long_description = fh.read()`

File: py_async_grpc_prometheus/prometheus_async_server_interceptor.py (hunk 3 above)
Comment (avik-taranis) [reply py_async_grpc_prometheus/prometheus_async_server_interceptor.py:154]: Agree - opening an issue for it.

**Pull request merged**